src/main/ruby/truffleruby/core/struct.rb

```diff
@@ -154,8 +155,30 @@ def initialize(*args)
       raise ArgumentError, "Expected #{attrs.size}, got #{args.size}"
     end
 
-    attrs.each_with_index do |attr, i|
-      TrufflePrimitive.object_hidden_var_set self, attr, args[i]
+    if self.class::KEYWORD_INIT
+      return if args.empty?
+      
+      if args.length > 1 || !args.first.is_a?(Hash)
+        raise ArgumentError, "wrong number of arguments (given #{args.size}, expected 0)"
+      end
+      kw_args = args.first
+
+      unknowns = []
+      kw_args.each_pair do |attr, value|
+        if attrs.include?(attr)
+          TrufflePrimitive.object_hidden_var_set self, attr, value
+        else
+          unknowns << attr
+        end
+      end
+
+      if unknowns.size > 0
```

Comment:
This would be a bit simpler as `unless unknowns.empty?`, I'll do that minor change.